kernel/sched.c

```diff
@@ -1285,39 +1285,44 @@ static void halt_thread(struct k_thread *thread, uint8_t new_state)
 	/* We hold the lock, and the thread is known not to be running
 	 * anywhere.
 	 */
-	if ((thread->base.thread_state & new_state) == 0U) {
+	if (likely((thread->base.thread_state & new_state) == 0U)) {
 		thread->base.thread_state |= new_state;
 		if (z_is_thread_queued(thread)) {
 			dequeue_thread(thread);
 		}
 
-		if (new_state == _THREAD_DEAD) {
-			if (thread->base.pended_on != NULL) {
-				unpend_thread_no_timeout(thread);
-			}
-			(void)z_abort_thread_timeout(thread);
-			unpend_all(&thread->join_queue);
+		if (likely(new_state == _THREAD_SUSPENDED)) {
```

Comment:
Seems like "suspend" is a rarer and more edgy case than "abort".  Though plausibly you could argue that it's more likely to be part of a performance path and thus should be the fast case for code generation.